### PR TITLE
python plugin: find setup.py when source-subdir is used

### PR DESCRIPTION
**Pull request by sergiusens**
LP: #1752481

Signed-off-by: Sergio Schvezov <sergio.schvezov@canonical.com>

- [x] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] If this is a bugfix. Have you checked that there is a bug report open for the issue you are trying to fix on [bug reports](https://bugs.launchpad.net/snapcraft)?
- [x] If this is a new feature. Have you discussed the design on the [forum](https://forum.snapcraft.io)?
- [x] Have you successfully run `./runtests.sh static`?
- [x] Have you successfully run `./runtests.sh unit`?

-----

### CODE REVIEW

**Comment by kyrofa** [comment]
I think this is obviously correct as far as taking the source_subdir into account to find the setup.py, so I'll leave a +1 on that aspect.

However, I think there's a larger issue here. Building should not be reaching back into the sourcedir, it should be using the builddir. The build step calls `_install_project` which then calls `_get_setup_py_dir`, `_get_constraints`, and `_get_requirements` all of which use the sourcedir. These functions are also used in pulling, so they need to understand the context in which they're called to return the appropriate path.

**Comment by kyrofa** [comment]
> Ideally, this would happen only on the base plugin.

Agreed, like we do for builddir.

File: snapcraft/plugins/python.py [hunk 1]
```diff
@@ -238,8 +242,11 @@ def _get_constraints(self):
             if isurl(self.options.constraints):
                 constraints = {self.options.constraints}
             else:
-                constraints = {os.path.join(
-                    self.sourcedir, self.options.constraints)}
+                constraints = {self._find_file(
+                    filename=self.options.constraints)}
+                if not constraints:
+                    # TODO snapcraft owned message
+                    raise FileNotFoundError(self.options.constraints)
```
Comment (kyrofa) [inline snapcraft/plugins/python.py]: This is new behavior, is it not? Why not define the error class here? Also would be nice to see tests covering these cases.

Comment (sergiusens) [reply snapcraft/plugins/python.py]: we don't have any place to add them for plugins, this is urgent and needs to come out today (that is why we have a TODO)

File: snapcraft/plugins/python.py [hunk 2]
```diff
@@ -217,18 +217,22 @@ def build(self):
             self._python_major_version, stage_dir=self.project.stage_dir,
             install_dir=self.installdir)
 
+    def _find_file(self, *, filename: str) -> str:
+        # source-subdir defaults to ''
+        for basepath in [self.builddir, self.sourcedir]:
```
Comment (kyrofa) [inline snapcraft/plugins/python.py]: Builddir includes the subdir, FYI.

File: snapcraft/plugins/python.py (hunk 1 above)
Comment (kyrofa) [reply snapcraft/plugins/python.py]: That [hasn't stopped us in the past](https://github.com/snapcore/snapcraft/blob/master/snapcraft/plugins/catkin.py#L103). Your call, of course. No tests either, then?

File: snapcraft/plugins/python.py (hunk 2 above)
Comment (sergiusens) [reply snapcraft/plugins/python.py]: yeah, that's in the last push :-/

Comment (kyrofa) [reply snapcraft/plugins/python.py]: Would using `self.build_basedir` not be cleaner than special-casing the sourcedir?

**Comment by sergiusens** [comment]
@kyrofa fwiw, this will not land in master but will be made available to outside consumers

**Pull request merged**